---
title: "Utilization of Tandem Repeats Finder (TRF)"
author: "Mira Sohn" 
output: html_document
---



### Aims

#### - This workflow is aimed at finding and counting repetitive elements based on ensembl gene id using the Tandem Repeats Finder (Unix/Linux ver.)

#### - References: [TRF Web](https://tandem.bu.edu/trf/trf.html), [TRF linux](https://tandem.bu.edu/trf/trf.unix.help.html), [TRF paper](https://pubmed.ncbi.nlm.nih.gov/9862982/), [biomaRt Doc](https://bioconductor.org/packages/release/bioc/html/biomaRt.html)


### Conda environment

#### - This analysis was performed under [conda](https://conda.io/projects/conda/en/latest/index.html) environment (see the [recipe](https://github.com/Mira0507/TRF_demo/blob/master/conda_r.yml))


```{r global_options, include=FALSE}


knitr::opts_chunk$set(
    warning=FALSE,
    message=FALSE
    )


```



### Loading packages



```{r loading_packages}

library(data.table)
library(tidyverse)
library(biomaRt)

```


### Importing gene ids 

#### - This demonstration uses ensembl human gene id 


```{r importing_geneid}

# Importing my genes to a vector 
geneid <- scan("geneid.txt", character(), quote="")

# Exploring the imported object 
geneid[1:10]
class(geneid)
length(geneid)



```


### Retrieving cDNA sequences by connecting to Ensembl with biomaRt

#### - This step uses useMart() and getSequence() functions from biomaRt 



```{r count_repetitive_elements}



# listEnsembl() <------ shows the BioMart databases hosted by Ensembl (e.g. snp)

# listDatasets(useMart('ensembl')) <--- shows available datasets 
# (e.g.hsapiense_gene_ensembl, hsapiense_snp)  





# Creating a connection to ensembl human gene db
ens.mart <- useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")

# Exploring Ensembl filters attributes 

# att <- listAttributes(ens.mart)
# fil <- listFilters(ens.mart)   # Returns a data frame


# att[1:20,]
# fil[1:20,]



# Creating an empty data frame
seqTable <- data.frame()


# Getting the longest cDNA sequences per gene id
# (This execution can take up to several mimutes!)
for (i in 1:length(geneid)) {

    # Assigning a gene id
    id <- geneid[i]

    # Getting sequences along with the length
    df <- getSequence(id=id, 
                      type="ensembl_gene_id", 
                      seqType="cdna", 
                      mart=ens.mart, 
                      verbose=F) %>%
            mutate(seqsize=nchar(cdna)) # This new column stores the size of each cDNA

    # Filtering the longest cDNA per GENEID
    # (Multiple cDNA (transcripts) are found in one gene id)
    df <- df[df$seqsize == max(df$seqsize), ] 

    seqTable <- rbind(seqTable, df)
}

# Explore the output
head(seqTable)
dim(seqTable)

```

### Running TRF

#### - This workflow uses Linux TRF 

#### - The input file format has to be **FASTA** (see below example)

#### - The output files are either html and/or .dat files depending on flags set 

#### - This workflow uses recommended parameters (see [Unix/linux TRF](https://tandem.bu.edu/trf/trf.unix.help.html))


```{r fasta_example}


# Below is an example of FASTA format
#
#
# >ENSG00000125148
# GGTTTATTTTCCAGATGCAATCAATGCCCC....
# >ENSG00000101361
# CACTTCTAGACCAAAACTGCAAAGGAAC......

```


```{r run_trf}

# Creating an input fasta file storing gene ids and cDNA sequences (as fasta format!!)
for (i in 1:length(geneid)) {

    # Adding a line for my gene id (e.g. >ENSG00000125148)
    line1 <- paste0(">", geneid[i]) 

    # Adding a line for cDNA sequences 
    line2 <- seqTable$cdna[i]


    write(line1, file="trf_input.fa", append=T)
    write(line2, file="trf_input.fa", append=T) # trf_input.fa is my input for trf

}


# Running tandem repeats finder (trf)
# Command: 
# trf <File> <Match> <Mismatch> <Delta> <PM> <PI> <Minscore> <MaxPeriod>
# -h: avoids to create html output
# -d: creates a .dat file 
system("trf trf_input.fa 2 7 7 80 10 50 500 -d -h")  # Returns a .dat file 

# Done.> pops up when it's done  
# "trf_input.fa.2.7.7.80.10.50.500.dat" (=output) is created in my working directory 



```


### Exploring analyzed repetitive elements 


```{r data_cleaning}




# Converting the TRF results from the .dat file to a data frame
# 
# 
# Isolating every single line of info to the trflines object
trflines <- readLines("trf_input.fa.2.7.7.80.10.50.500.dat")  

# Exploring the trflines object
class(trflines)
length(trflines)
trflines[1:20]

# Creating an empty data frame
trfTable <- data.frame()

# Storing every single line to the data frame except empty ("") and 
# unnecessary (1 to 6) lines 
# (The line 1 to 6 of trflines have title & description about this run)
for (i in 7:length(trflines)) {

    if (trflines[i] != "") {


        df <- data.frame(Line=trflines[i])

        trfTable <- rbind(trfTable, df)

    }



}


# Exploring the output data frame
head(trfTable)
dim(trfTable)
class(trfTable)


# Clean the TRF data frame 
# (removing rows containing parameters and adding line codes)
trfTable <- trfTable %>%
    mutate(code=ifelse(grepl("Sequence", Line, fixed=T), "S", 
                       ifelse(grepl("Parameters", Line, fixed=T), "P",
                       "CDNA"))) %>% dplyr::filter(code != "P")


# Exploring the output data frame
head(trfTable)
dim(trfTable)


# Extracting the number of repetitive elements per GENEID
#
# Creating an empty data frame 
trfnumTable <- data.frame()

# Converting the line code to a character vector
code.vector <- trfTable$code


# Creating empty vectors 
GENE <- c()    
COUNT <- c()

# Creating variables for storing counts 
num.s <- 0
num.c <- 0

for (i in 1:nrow(trfTable)) {

    # If the line code is "S" (=Sequence: gene id)
    if (code.vector[i] == "S") {

        num.s <- num.s + 1
        num.c <- 0

    # If the line code is "CDNA" (=Repeats) 
    } else {

        num.s <- 0 
        num.c <- num.c + 1
    }

    GENE[i] <- num.s # Indicates gene id coded "S"
    COUNT[i] <- num.c  # Indicates repetitive elements coded "CDNA" 


}


# Exploring the GENE and Count vectors
GENE
COUNT

# GENE[i] = 0: repetitive element line 
# GENE[i] > 0: gene id line
#   1  2  3  0  1  2  3  4  0  0  1
#         |  |           |  |  |
#     geneid |       geneid |  |
#           repeat1     repeat1|
#                            repeat2
#
#
#
# COUNT[i] = 0: gene id line
# COUNT[i] > 0: repetitive element line 
#   0  0  0  1  0  0  0  0  1  2  0 
#         |  |           |  |  |  
#     geneid |       geneid |  |
#          repeat1      repeat1|
#                           repeat2


# Assign vectors storing line indices of 
# gene id (repetitive.geneid), the first repetitive elements (repetitive.first), 
# and the last repetitive elements (repetitive.last)
repetitive.geneid <- which(COUNT == 1) - 1  # right before the first repetitive elements
repetitive.first <- which(COUNT == 1)  # the first repetitive elements 
repetitive.last <- which(GENE == 1) - 1  # the last repetitive elements per gene id


# Exploring the indices
repetitive.geneid
repetitive.first
repetitive.last


# In case that the first and/or the last gene had one repetitive element: 
if (repetitive.last[1] < repetitive.first[1]) {

    repetitive.last <- repetitive.last[-1]

    if (repetitive.last[length(repetitive.last)] < 
        repetitive.first[length(repetitive.first)]) {

        repetitive.last <- c(repetitive.last, 
                             repetitive.first[length(repetitive.last)])

    }



    

}

# Explore the indices (all three vectors have to have the same length!)
repetitive.geneid
repetitive.first
repetitive.last



# Summarize the gene id (GENEID), indices of the first (Repeats_First) and 
# the last (Repeats_Last) repetitive elements in each gene id
trfnumTable <- data.frame(GENEID=trfTable$Line[repetitive.geneid],  
                          Repeats_First=repetitive.first, 
                          Repeats_Last=repetitive.last) %>% mutate(Repeats_Count=Repeats_Last - Repeats_First + 1,
                                   GENEID=str_replace(GENEID, "Sequence: ", "")) 


# Exploring the output data frame
head(trfnumTable)
dim(trfnumTable)



# Explore the output data frame
glimpse(trfnumTable)

# Export as a csv file
write.csv(trfnumTable, "TRF_demo.csv")
```




### Session Info

```{r session_info}

sessionInfo()

```

